feat: Implement IR cursor to allow searches over IR streams.

anlowee opened the pull request:
<!-- markdownlint-disable MD012 -->

<!--
Set the PR title to a meaningful commit message that:

* is in imperative form.
* follows the Conventional Commits specification (https://www.conventionalcommits.org).
  * See https://github.com/commitizen/conventional-commit-types/blob/master/index.json for possible
    types.

Example:

fix: Don't add implicit wildcards ('*') at the beginning and the end of a query (fixes #390).
-->

# Description

<!-- Describe what this request will change/fix and provide any details necessary for reviewers. -->

~This PR is the successor of #25 , and must be merged after #25 is merged.~(aleady merged)

This PR added the implementation of the `BaseClpCursor` for IR to support search on IR stream. Currently since the auto-generated keys are queried by a leading `@` in the field name, the Velox doesn't support this special char now. We will figure out how to support query on auto-generated keys in the future. In this PR, we only consider the user-generated keys.

This PR doesn't populate the timestamp type yet, will support in the next PR.

This PR doesn't modify the docs, will add in the next PR.

# Checklist

<!-- Ensure each item below is satisfied and indicate so by inserting an `x` within each `[ ]`. -->

* [x] The PR satisfies the [contribution guidelines][yscope-contrib-guidelines].
* [x] This is a breaking change and that has been indicated in the PR title, OR this isn't a
  breaking change.
* [x] Necessary docs have been updated, OR no docs need to be updated.

# Validation performed

<!-- Describe what tests and validation you performed on the change. -->

Passed the CI.
End-to-end test.

[yscope-contrib-guidelines]: https://docs.yscope.com/dev-guide/contrib-guides-overview.html


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- New Features
  - Added IR-based streaming cursor with projection-aware lazy loading and support for archive and IR inputs; improved Filesystem and S3 handling.

- Refactor
  - Unified cursor and split handling, simplified vector creation and query execution paths.

- Tests
  - Added IR-output verification for multiple pushdown and non-pushdown scenarios.

- Chores
  - Updated external CLP dependency tag and extended include search paths for the CLP helper library.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

anlowee committed:
Extract ClpCursor abstract class and move the original ClpCursor implementation to ClpArchiveCursor

anlowee committed:
Merge the code

anlowee committed:
Rename the extracrted ClpCursor to BaseClpCursor, and add the skeleton of ClpIrCursor

anlowee committed:
WIP: implement the ClpIrCursor, one todo is to marshal the row in ClpVectorLoader

anlowee committed:
Remove the IR related code and add a subpackage to store archive search related code

anlowee committed:
Address coderabbitai comments

anlowee committed:
Address comments

anlowee committed:
Update from part1 branch

anlowee committed:
WIP

anlowee committed:
Move the current VectorLoader to Archive-specific

anlowee committed:
Merge branch 'xwei/ir-integration-phase2-part1' into xwei/ir-integration-phase2-part2

anlowee committed:
Rename and WIP

anlowee committed:
Put more archive specific thing into Archive code

anlowee committed:
Merge branch 'xwei/ir-integration-phase2-part1' into xwei/ir-integration-phase2-part2

anlowee committed:
WIP

anlowee committed:
WIP

anlowee committed:
Merge branch 'presto-0.293-clp-connector' into xwei/ir-integration-phase2-part1

anlowee committed:
Merge branch 'xwei/ir-integration-phase2-part1' into xwei/ir-integration-phase2-part2

anlowee committed:
Get it work

anlowee committed:
Code clean up

anlowee committed:
Merge branch 'xwei/ir-integration-phase2-part1' into xwei/ir-integration-phase2-part2

anlowee committed:
Address coderabbitai comments

anlowee committed:
Merge branch 'xwei/ir-integration-phase2-part1' into xwei/ir-integration-phase2-part2

anlowee committed:
Minor fix

anlowee committed:
Fix unitest

anlowee commented:
@coderabbitai Please just review the code under velox/connectors/clp/search_lib/ir and unittests.

anlowee committed:
Fix building issues

anlowee committed:
Address comments

anlowee committed:
Merge branch 'xwei/ir-integration-phase2-part1' into xwei/ir-integration-phase2-part2

anlowee committed:
Try out zhihao fix

anlowee committed:
Apply Zhihao fix

anlowee committed:
Merge branch 'presto-0.293-clp-connector' into xwei/ir-integration-phase2-part2

anlowee committed:
Address some coderabbitai comments after merging the first PR

gibber9809 commented on velox/connectors/clp/search_lib/ir/CMakeLists.txt:
```diff
@@ -0,0 +1,32 @@
+  clp-s-ir-search
+  PUBLIC clp_s::archive_reader
+  PRIVATE
+    clp_s::clp_dependencies
```
We should make a note to export library targets for the IR stream code inside of the CLP repo -- relying on `clp_s::clp_dependencies` will work for now, but isn't ideal long term.

anlowee replied:
How about I adding a comment "# After IR-stream-related targets are exported, we should replace this with finer granularity dependencies" above `clp_s::clp_dependencies`

anlowee replied:
`# Once the IR-stream-related targets are exported, this should be updated to use more fine-grained dependencies.`

gibber9809 commented on velox/connectors/clp/search_lib/ir/ClpIrUnitHandler.cpp:
```diff
@@ -0,0 +1,43 @@
+
+void ClpIrUnitHandler::clearFilteredLogEvents() {
+  filteredLogEvents_->clear();
+}
```
These functions seem pretty short. Maybe we could just leave all the implementations in the header?

anlowee replied:
It seems these are the only three functions. Shall we only keep the `.h` file after moving the implementation to the header?

anlowee replied:
Deleted `.cpp` file.

gibber9809 commented on velox/connectors/clp/search_lib/ir/ClpIrCursor.h:
```diff
@@ -0,0 +1,89 @@
+                  projected_key_and_index)
+      -> ystdlib::error_handling::Result<void> {
+    projectedColumnNameNodeIdMap_.insert(
+        {std::string(projected_key_and_index.first), nodeId});
```
Now that this callback also gives the index of the resolved projection within the original list of projections I think we should use that instead of the name -- it will allow this code to work correctly for cases where there are duplicate columns in the projection, and when there are two columns with the same name but different projected type.

So basically change `projectedColumnNameNodeIdMap` to a `std::unordered_map<size_t, ::clp::ffi::SchemaTree::Node::id_t` and store the mapping between `projected_key_and_index.second` and `nodeId`. Then later when we're checking what each projected column maps to instead of going `index` -> `name` -> `projected nodes` we can just directly look up `index` -> `projected nodes`.

gibber9809 commented on velox/connectors/clp/search_lib/ir/ClpIrCursor.cpp:
```diff
@@ -0,0 +1,208 @@
+      "Reader index out of bounds");
+  auto projectedColumn = outputColumns_[readerIndex_];
+  auto projectedColumnType = projectedColumn.type;
+  auto it = projectedColumnNameNodeIdMap_.find(projectedColumn.name);
```
Related to my other comment -- we should end up being able to directly look up `projectedColumnIndexNodeIdMap_.find(readerIndex_)`.

gibber9809 commented on velox/connectors/clp/search_lib/ir/ClpIrCursor.cpp:
```diff
@@ -0,0 +1,208 @@
+  vector->setNulls(allocateNulls(vectorSize, pool, bits::kNull));
+  VELOX_CHECK_LT(
+      readerIndex_,
+      projectedColumnNameNodeIdMap_.size(),
```
```suggestion
      outputColumns_.size(),
```
Otherwise we would crash in situations where not all columns have been resolved yet when we should just be marshalling nulls for unresolved fields.

gibber9809 commented on velox/connectors/clp/search_lib/ir/ClpIrCursor.cpp:
```diff
@@ -0,0 +1,208 @@
+  VELOX_CHECK(
+      it != projectedColumnNameNodeIdMap_.end(),
+      "Projected column '{}' not found in node id map",
+      projectedColumn.name);
```
Related to my other comment -- instead of asserting here we should be marshalling nulls for fields that haven't resolved to any specific nodes.

gibber9809 commented on velox/connectors/clp/search_lib/ir/ClpIrCursor.cpp:
```diff
@@ -0,0 +1,208 @@
+      }
+      return error;
+    }
+    cnt++;
```
I think we're trying to count matching results here right? In that case we need to rely on the fact that for search the `handle_log_event` method on the `IrUnitHandler` is only invoked for matching results. This current code will actually count for every single IR unit which will include non-matching log events, new schema tree nodes, etc..

I think we should be able to do something like `whille(ir_unit_handler.get_num_filtered_log_events() < numRows)` instead.

anlowee replied:
Here actually we are counting the scanned log events. The `numRows` is the current batch size, the `cnt` here is to ensure it doesn't scan more than batch size. The number of matching results returns by `getNumFilteredRows` function above.

Do you think we change the `filtered` to `matching`? The `filtered` is inherited from previous code.

gibber9809 replied:
I see, in that case what we should do is something like
```
if (clp::ffi::ir_stream::IrUnitType::LogEvent == deserializeResult.value()) {
  ++cnt;
}
```
to make sure we're just counting scanned log events instead of counting _all_ IR units.

gibber9809 commented on velox/connectors/clp/search_lib/ir/ClpIrCursor.cpp:
```diff
@@ -0,0 +1,208 @@
+  while (cnt < numRows) {
+    auto deserializeResult =
+        irDeserializer_->deserialize_next_ir_unit(*irReader_);
+    if (deserializeResult.has_error()) {
+      auto error = deserializeResult.error();
+      if (std::errc::result_out_of_range == error ||
+          irDeserializer_->is_stream_completed()) {
+        break;
+      }
+      return error;
```
Judging by some comments in the clp code it looks like we might switch from `result_out_of_range` to another error code to indicate truncated IR streams in the future -- we should probably put up an issue or something to track that so that we remember to change this code at the same time.

anlowee replied:
@LinZhihao-723 Do we have such an issue?

LinZhihao-723 replied:
Yes. We were planning to add our own error code for IR serde. I think we will have it in at some point.

anlowee replied:
Do we have an issue link so that I can put it in the code comment?

gibber9809 reviewed (changes requested):
Partial review for code under the `ir` subdirectory. Still need to review `ClpIrVectorLoader.cpp` and `ClpIrVectorLoader.h`, but those two files should be more straightforward.

gibber9809 commented on velox/connectors/clp/search_lib/ir/ClpIrVectorLoader.h:
```diff
@@ -0,0 +1,57 @@
+
+ private:
+  inline static thread_local std::unique_ptr<simdjson::ondemand::parser>
+      arrayParser_ = std::make_unique<simdjson::ondemand::parser>();
```
Does this need to be a `unique_ptr`? Also, instead of having this be a static thread_local member maybe we could just have it be a regular member variable in `ClpIrCursor` and pass it down for marshalling? Having this as a static thread local variable makes the code harder to reason about, especially when debugging errors.

gibber9809 replied:
I guess since this follows the `LazyVector` model it might be a bit difficult to have it be a `ClpIRCursor` member. Even still I think the code is easier to interpret if either
1. We just directly make this member variable `simdjson::ondemand::parser arrayParser_` or
2. We wrap it in optional `std::optional<simdjson::ondemand::parser>` and initialize it only when needed if we're that concerned about the performance overhead of initializing/allocating buffers for the parser.

anlowee replied:
Nice catch. This was copied from the `ClpArchiveVectorLoader`, didn't notice this previously.

gibber9809 commented on velox/connectors/clp/search_lib/ir/ClpIrCursor.cpp:
```diff
@@ -0,0 +1,208 @@
+        literalType = search::ast::LiteralType::IntegerT;
+        break;
+      case ColumnType::String:
+        literalType = search::ast::LiteralType::VarStringT;
```
```suggestion
        literalType = search::ast::LiteralType::VarStringT | search::ast::LiteralType::ClpStringT;
```

gibber9809 commented on velox/connectors/clp/search_lib/ir/ClpIrVectorLoader.cpp:
```diff
@@ -0,0 +1,133 @@
+    auto value = userGenNodeIdValueMap.at(nodeId_);
+    if (!value.has_value()) {
+      continue;
+    }
```
```suggestion
    auto const value_it{userGenNodeIdValueMap.find(nodeId_)};
    if (userGenNodeIdValueMap.end() == value_it || false == value_it->has_value()) {
      continue;
    }
    auto const& value{*value_it};
```
Maybe something more like this? Avoids doing the lookup twice + makes the special case handling a bit easier to follow imo.

gibber9809 commented on velox/connectors/clp/search_lib/ir/ClpIrVectorLoader.cpp:
```diff
@@ -0,0 +1,133 @@
+        arrayVector->setOffsetAndSize(vectorIndex, 0ULL, numElements);
+        arrayVector->setNull(vectorIndex, false);
+        break;
+      }
```
I don't think rabbit is correct about the second part, but it's probably worth double checking if we're using `setOffsetAndSize` correctly.

gibber9809 reviewed (changes requested):
Partial review for the `ir` subdirectory, covering the code I did not look at in my first partial review.

anlowee committed:
Address comments

anlowee commented:
Addressed the comments of the first round review. Added the IR stream file for the `test_2` test case and unit test cases for testing marshalling array.

anlowee committed:
Address a comment

gibber9809 reviewed (approved):
For code in the IR subdirectory: LGTM!
